Backported from:

From: Stefan Floeren <42731906+stefan-floeren@users.noreply.github.com>
Date: Tue, 16 Apr 2019 08:38:01 +0200
Subject: [PATCH] Replace make_shared with new in some cases

Replace make_shared for asio types that take a lib::ref as a parameter.
This should fix the ASIO change (boostorg/asio@59066d8) for 1.70,
while keeping it backwards compatible to older boost versions.
---
 websocketpp/transport/asio/connection.hpp    | 7 ++++---
 websocketpp/transport/asio/endpoint.hpp      | 3 +--
 websocketpp/transport/asio/security/none.hpp | 3 +--
 websocketpp/transport/asio/security/tls.hpp  | 3 +--
 4 files changed, 7 insertions(+), 9 deletions(-)

diff --git a/websocketpp/transport/asio/connection.hpp b/websocketpp/transport/asio/connection.hpp
index 60f88a79..1ccda8f3 100644
--- a/websocketpp/transport/asio/connection.hpp
+++ b/websocketpp/transport/asio/connection.hpp
@@ -311,9 +311,10 @@ class connection : public config::socket_type::socket_con_type {
      * needed.
      */
     timer_ptr set_timer(long duration, timer_handler callback) {
-        timer_ptr new_timer = lib::make_shared<boost::asio::deadline_timer>(
-            lib::ref(*m_io_service),
-            boost::posix_time::milliseconds(duration)
+        timer_ptr new_timer(
+            new boost::asio::deadline_timer(
+                *m_io_service,
+                boost::posix_time::milliseconds(duration))
         );
 
         if (config::enable_multithreading) {
diff --git a/websocketpp/transport/asio/endpoint.hpp b/websocketpp/transport/asio/endpoint.hpp
index ddab2c74..4b719a97 100644
--- a/websocketpp/transport/asio/endpoint.hpp
+++ b/websocketpp/transport/asio/endpoint.hpp
@@ -195,8 +195,7 @@ class endpoint : public config::socket_type {
 
         m_io_service = ptr;
         m_external_io_service = true;
-        m_acceptor = lib::make_shared<boost::asio::ip::tcp::acceptor>(
-            lib::ref(*m_io_service));
+        m_acceptor.reset(new boost::asio::ip::tcp::acceptor(*m_io_service));
 
         m_state = READY;
         ec = lib::error_code();
diff --git a/websocketpp/transport/asio/security/none.hpp b/websocketpp/transport/asio/security/none.hpp
index 5c8293db..6c7d3524 100644
--- a/websocketpp/transport/asio/security/none.hpp
+++ b/websocketpp/transport/asio/security/none.hpp
@@ -168,8 +168,7 @@ class connection : public lib::enable_shared_from_this<connection> {
             return socket::make_error_code(socket::error::invalid_state);
         }
 
-        m_socket = lib::make_shared<boost::asio::ip::tcp::socket>(
-            lib::ref(*service));
+        m_socket.reset(new boost::asio::ip::tcp::socket(*service));
 
         m_state = READY;
 
diff --git a/websocketpp/transport/asio/security/tls.hpp b/websocketpp/transport/asio/security/tls.hpp
index c76fd9aa..04ac3790 100644
--- a/websocketpp/transport/asio/security/tls.hpp
+++ b/websocketpp/transport/asio/security/tls.hpp
@@ -193,8 +193,7 @@ class connection : public lib::enable_shared_from_this<connection> {
         if (!m_context) {
             return socket::make_error_code(socket::error::invalid_tls_context);
         }
-        m_socket = lib::make_shared<socket_type>(
-            _WEBSOCKETPP_REF(*service),lib::ref(*m_context));
+        m_socket.reset(new socket_type(*service, *m_context));
 
         m_io_service = service;
         m_strand = strand;
